vector/src/main/java/fr/gouv/tchap/activity/TchapRoomAccessByLinkActivity.kt

```diff
@@ -0,0 +1,364 @@
+        if (null != powerLevels) {
+            val powerLevel = powerLevels.getUserPowerLevel(session.myUserId)
+            isAdmin = powerLevel >= CommonActivityUtils.UTILS_POWER_LEVEL_ADMIN
+        }
+
+        if (isAdmin && isConnected && isForum == false) {
+            roomAccessByLinkStatus.visibility = View.GONE
+            switchRoomAccessByLink.visibility = View.VISIBLE
+        } else {
+            roomAccessByLinkStatus.visibility = View.VISIBLE
+            switchRoomAccessByLink.visibility = View.GONE
+        }
+
+        switchRoomAccessByLink.isEnabled = false
+        if (RoomState.JOIN_RULE_INVITE == joinRule) {
+            roomAccessByLinkStatus.text = getString(R.string.tchap_room_settings_room_access_by_link_disabled)
+            switchRoomAccessByLink.isChecked = false
+            roomAccessByLinkInfo.text = getString(R.string.tchap_room_settings_enable_room_access_by_link_info_off)
+            roomAccessLink.visibility = View.GONE
+            buttonsContainer.visibility = View.GONE
+        } else {
+            roomAccessByLinkStatus.text = getString(R.string.tchap_room_settings_room_access_by_link_enabled)
+            switchRoomAccessByLink.isChecked = true
+            roomAccessByLinkInfo.text = getString(R.string.tchap_room_settings_enable_room_access_by_link_info_on)
+            roomAccessLink.text = createPermalink(getString(R.string.permalink_prefix), room.state.canonicalAlias)
+            roomAccessLink.visibility = View.VISIBLE
+            buttonsContainer.visibility = View.VISIBLE
+        }
+        switchRoomAccessByLink.isEnabled = true
+    }
+
+    private fun getRoomDirectoryVisibility() {
+        showWaitingView()
+        room.getDirectoryVisibility(room.roomId, object: ApiCallback<String> {
+            override fun onSuccess(visibility: String) {
+                isForum = RoomDirectoryVisibility.DIRECTORY_VISIBILITY_PUBLIC == visibility
+                hideWaitingView()
+                refreshDisplay()
+            }
+
+            private fun onError(errorMessage: String) {
+                Log.e(LOG_TAG, "## getRoomDirectoryVisibility: failed $errorMessage")
+                // We consider here the room as a forum in order to prevent edition of the option
+                // Indeed the switch is hidden in case of a forum room
+                isForum = true
+                hideWaitingView()
+                refreshDisplay()
+            }
+
+            override fun onNetworkError(e: Exception) {
+                onError(e.localizedMessage)
+            }
+
+            override fun onMatrixError(e: MatrixError) {
+                onError(e.message)
+            }
+
+            override fun onUnexpectedError(e: Exception) {
+                onError(e.localizedMessage)
+            }
+        })
+    }
+
+    private fun forbidGuestAccess(callback: ApiCallback<Void>) {
+        if (room.state.guestAccess == RoomState.GUEST_ACCESS_FORBIDDEN) {
+            callback.onSuccess(null)
```

Comment:
Or I think this code could crash

Reply:
During my tests, no crash was observed for this code.
If I replace `Void` with `Unit`, this makes trouble to use `callback` in `room.updateGuestAccess()` call

Reply:
Please check at other places when we are using `Void` then.

Reply:
both `<Void>`and <Void?> are used in the existing code. I would vote for `<Void?>